core/emulate.c

```diff
@@ -920,17 +987,63 @@ static void em_movsx(struct em_context_t *ctxt)
     ctxt->dst.value = 0;
     memcpy(&ctxt->dst.value, &extension, ctxt->dst.size);
     memcpy(&ctxt->dst.value, &ctxt->src1.value, ctxt->src1.size);
+    return EM_CONTINUE;
+}
+
+static em_status_t em_push(struct em_context_t *ctxt)
+{
+    em_status_t rc;
+    em_operand_t sp;
+
+    // Pre-decrement stack pointer
+    register_add(ctxt, REG_RSP, -(int64_t)ctxt->operand_size);
+
+    memset(&sp, 0, sizeof(sp));
+    sp.type = OP_MEM;
+    sp.size = ctxt->operand_size;
+    sp.mem.ea = gpr_read(ctxt, REG_RSP, ctxt->address_size);
+    sp.mem.seg = SEG_SS;
+    sp.value = ctxt->src1.value;
+
+    rc = operand_write(ctxt, &sp);
```

Comment:
Reusing `operand_write()` to emulate the memory copy is really neat! There's just a little problem: the temporary `sp` operand is not preserved across re-entry of `em_emulate_insn()`, so we'll run into an infinite loop of pending MMIO writes if we are to emulate a push to MMIO (although that's very unlikely). How about just adding the implicit destination operand (`sp`) to `em_context_t`?

Reply:
Good idea, thank you! I've done that, although I've reused existing slots, i.e. `ctxt->dst` for *em_push*, and `ctxt->src1` for *em_pop*.